homeassistant/components/pyload/coordinator.py

```diff
@@ -0,0 +1,73 @@
+"""Update coordinator for pyLoad Integration."""
+
+from datetime import timedelta
+import logging
+
+from pyloadapi import (
+    CannotConnect,
+    InvalidAuth,
+    ParserError,
+    PyLoadAPI,
+    StatusServerResponse,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryError
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=20)
+
+
+class pyLoadData(StatusServerResponse):
+    """Data from pyLoad."""
+
+    free_space: int
+    version: str
+
+
+class PyLoadCoordinator(DataUpdateCoordinator[pyLoadData]):
+    """pyLoad coordinator."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, pyload: PyLoadAPI) -> None:
+        """Initialize pyLoad coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.pyload = pyload
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        try:
+            free_space = await self.pyload.free_space()
+            status = await self.pyload.get_status()
+            version = await self.pyload.version()
+        except InvalidAuth:
+            _LOGGER.debug("Authentication failed, trying to reauthenticate")
+            try:
+                await self.pyload.login()
+            except InvalidAuth as e:
+                raise ConfigEntryError(
+                    f"Authentication failed for {self.pyload.username}, check your login credentials",
+                ) from e
+            else:
+                raise UpdateFailed(
+                    "Unable to retrieve data due to cookie expiration but re-authentication was successful."
+                )
```

Comment:
trying to login in an except is meh tbf

Reply:
How would you trigger a re-login?

Reply:
I mean, if `InvalidAuth` is raised, is that always because the username/password is incorrect or is there another reason why we want this here

Reply:
The cookie has an expiration time and has to be refreshed. `get_status` can also return an `InvalidAuth `but raising a `ConfigEntryError` if we can recover, doesn't seem quite right. Does `ConfigEntryAuthFailed` trigger a re-login?

Reply:
Yes. Is there a way we can check the cookie before requesting?

Reply:
Not that i know of. Also the cookie gets invalidated when pyLoad is restartet, so tracking the expiration time is not an option and doing a login before every request would fill up pyLoads logfile pretty quick.

Reply:
maybe then have a boolean in the coordiantor to tell if we already tried logging in, and then just do return self._async_update_data, and if the bool is true, raise ConfigEntryError

Reply:
Here i'm doing something similar, only that it is a token refresh:

https://github.com/home-assistant/core/blob/473b3b61ebea3096ec75ebcf54c25615a32f3247/homeassistant/components/bring/coordinator.py#L53-L69

Reply:
Hmm, I am not sure if I like this tbf

Reply:
Can we agree to leave it for now and change it with the reauth flow?